prettyprinter/src/Data/Text/Prettyprint/Doc/Internal.hs

```diff
@@ -1785,7 +1794,16 @@ layoutWadlerLeijen
       | fits pWidth minNestingLevel availableWidth x = x
       | otherwise = y
       where
-        minNestingLevel = min lineIndent currentColumn
+        minNestingLevel =
+            -- See https://github.com/quchen/prettyprinter/issues/83.
+            if startsWithLine y
+                -- y might be a (more compact) hanging layout. Let's check x
+                -- thoroughly with the smaller lineIndent.
+                then lineIndent
+                -- y definitely isn't a hanging layout. Let's allow the first
+                -- line of x to be checked on its own and format it consistently
+                -- with subsequent lines with the same indentation.
+                else currentColumn
```

Comment:
These are the general conditions under which `layoutSmart`, given a `Union x y`, will return `x` instead of `y` after this change:

* `y` doesn't start with a line break
* `lineIndent` must be `< currentColumn`
* a line of `x` (but not the first) with an indentation of `lineIndent < i <= currentColumn` is too wide.

Reply:
> * a line of `x` (but not the first) with an indentation of `lineIndent < i <= currentColumn` is too wide.

Actually that line must be "reachable" too. I.e. there's no other line with an indentation `<= lineIndent` before it.